File: 백준/Gold/12869. 뮤탈리스크/뮤탈리스크.java
import java.util.StringTokenizer;
import java.io.*;
import java.util.Arrays;
public class Main {
    static int N;
    static int[] SVC;
    static int[][][] dp;
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        N = Integer.parseInt(br.readLine());
        SVC = new int[3];
        StringTokenizer st = new StringTokenizer(br.readLine());
        for (int i = 0; i < N; i++) {
            SVC[i] = Integer.parseInt(st.nextToken());
        }
        dp = new int[61][61][61];
        for (int i = 0; i < 61; i++) {
            for (int j = 0; j < 61; j++) {
                Arrays.fill(dp[i][j], -1);
            }
        }

        System.out.println(dfs(SVC[0], SVC[1], SVC[2]));
    }
    static int dfs(int a, int b, int c) {
        a = Math.max(a, 0);
        b = Math.max(b, 0);
        c = Math.max(c, 0);

        if (a == 0 && b == 0 && c == 0) {
            return 0;
        }

        if (dp[a][b][c] != -1) {
            return dp[a][b][c];
        }

        int result = 987654321;

        result = Math.min(result, dfs(a-9, b-3, c-1) + 1);
        result = Math.min(result, dfs(a-9, b-1, c-3) + 1);
        result = Math.min(result, dfs(a-3, b-9, c-1) + 1);
        result = Math.min(result, dfs(a-3, b-1, c-9) + 1);
        result = Math.min(result, dfs(a-1, b-9, c-3) + 1);
        result = Math.min(result, dfs(a-1, b-3, c-9) + 1);

        return dp[a][b][c] = result;
    }
}

File: 백준/Gold/12869. 뮤탈리스크/뮤탈리스크.py
import sys
sys.setrecursionlimit(10**6)

N = int(input())
SCV = list(map(int, input().split()))
for _ in range(3-N):
    SCV.append(0)

dp = {}

def dfs(a, b, c):
    # 재방문인 경우
    if (a, b, c) in dp:
        return dp[(a, b, c)]

    # 탈출조건
    if (a, b, c) == (0, 0, 0):
        return 0

    answer = min(dfs(max(a - 9, 0), max(b - 3, 0), max(c - 1, 0)),
                 dfs(max(a - 9, 0), max(b - 1, 0), max(c - 3, 0)),
                 dfs(max(a - 3, 0), max(b - 9, 0), max(c - 1, 0)),
                 dfs(max(a - 3, 0), max(b - 1, 0), max(c - 9, 0)),
                 dfs(max(a - 1, 0), max(b - 3, 0), max(c - 9, 0)),
                 dfs(max(a - 1, 0), max(b - 9, 0), max(c - 3, 0))) + 1

    dp[(a, b, c)] = answer
    return answer

print(dfs(*SCV))